src/AI/Muriel.java

```diff
@@ -0,0 +1,750 @@
+package AI;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Queue;
+import java.util.Set;
+
+import CommandingOfficers.Commander;
+import CommandingOfficers.CommanderAbility;
+import Engine.GameAction;
+import Engine.GameAction.ActionType;
+import Engine.GameActionSet;
+import Engine.Utils;
+import Engine.XYCoord;
+import Engine.Combat.BattleInstance;
+import Terrain.Environment;
+import Terrain.Environment.Weathers;
+import Terrain.GameMap;
+import Terrain.Location;
+import Terrain.TerrainType;
+import Units.Unit;
+import Units.UnitModel;
+import Units.Weapons.Weapon;
+
+/**
+ * Muriel will Make Units Reactively, Informed by the Enemy Loadout.
+ */
+public class Muriel implements AIController
+{
+  private Queue<GameAction> queuedActions = new ArrayDeque<GameAction>();
+
+  private Commander myCo = null;
+
+  private StringBuffer logger = new StringBuffer();
+  private int turnNum = 0;
+
+  private ArrayList<Commander> enemyCos = null;
+  
+  private UnitEffectivenessMap myUnitEffectMap;
+  private final double COST_EFFECTIVENESS_THRESHOLD = 0.75;
+
+  private ArrayList<XYCoord> nonAlliedProperties; // set from AIUtils.
+
+  public Muriel(Commander co, Commander[] allCos )
+  {
+    myCo = co;
+
+    // Initialize UnitModel collections.
+    ArrayList<UnitModel> myUnitModels = new ArrayList<UnitModel>();
+    enemyCos = new ArrayList<Commander>();
+    Map<Commander, ArrayList<UnitModel> > otherUnitModels = new HashMap<Commander, ArrayList<UnitModel> >();
+    for( Commander other : allCos )
+    {
+      if( myCo.isEnemy(other) )
+      {
+        enemyCos.add(other);
+        otherUnitModels.put(other, new ArrayList<UnitModel>());
+      }
+    }
+
+    // Figure out what I and everyone else can build.
+    for( UnitModel.UnitEnum modelEnum : UnitModel.UnitEnum.values() )
```

Comment:
Why do you have this convoluted loop when CO.unitModels is public?

Reply:
I was initially planning to sort out only the direct-fire ones, but then reversed that decision and didn't rewrite the loops. Silly me.